### PR TITLE
[FEA] distributed gnn primitives

### PR DESCRIPTION
**Pull request by stadlmax**
# Modulus Pull Request

## Description
- note: hold back until #105 is merged, this PR builds upon these changes
- based on distributed primitives in #105, this PR adds distributed GNN primitives using these generic routines
- add a naive graph partitioning scheme
- modifies GraphCastNet and MeshGraphNet to (more or less) allow straightforward tensor-parallel distribution of these models
- adds tests for distributed runs for these two models testing underlying GNN primitives implicitly 

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/modulus/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [x] The [CHANGELOG.md](https://github.com/NVIDIA/modulus/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/modulus/issues) is linked to this pull request.

### CODE REVIEW

File: modulus/distributed/utils.py
Comment (stadlmax) [inline modulus/distributed/utils.py:620]: bugfix

**Comment by mnabian** [comment]
@stadlmax the coverage test is failing because it falls below 80%. Could you please use `# no pragma` in your code until we stand up the multi-GPU CI and coverage? Example: https://github.com/NVIDIA/modulus/blob/main/modulus/models/sfno/sfnonet.py#L104

File: modulus/models/graphcast/graph_cast_net.py
Comment (akshaysubr) [inline modulus/models/graphcast/graph_cast_net.py:134]: When this is `True`, how would one account for that in the dataloader? Is there a mechanism for that? If not, we should think about how we can implement some such mechanisms in a future PR.

File: modulus/models/graphcast/graph_cast_net.py
Comment (akshaysubr) [inline modulus/models/graphcast/graph_cast_net.py:174]: Debug code?

File: modulus/models/graphcast/graph_cast_net.py
Comment (akshaysubr) [inline modulus/models/graphcast/graph_cast_net.py:202]: Since this is a specific test in the try block, would be good to either catch only that specific exception so other exceptions still get raised or use something like `if hasattr(self.m2g_graph, "adj_tensors")`

File: modulus/models/graphcast/graph_cast_net.py
Comment (akshaysubr) [inline modulus/models/graphcast/graph_cast_net.py:227]: Same exception specification comment as above

File: modulus/models/layers/fused_silu.py
Comment (akshaysubr) [inline modulus/models/layers/fused_silu.py:22]: This is to accommodate earlier versions of PyTorch? Would also be good to specify an `ImportError` exception here

File: test/models/graphcast/test_graphcast_snmg.py
Comment (akshaysubr) [inline test/models/graphcast/test_graphcast_snmg.py:154]: Would it be worth to parameterize on checkpointing as well to ensure it works in conjunction with distributed graphs?

File: modulus/models/graphcast/graph_cast_net.py
Comment (stadlmax) [reply modulus/models/graphcast/graph_cast_net.py:251]: my mistake, left over from stupid printf debugging ^^

File: modulus/models/layers/fused_silu.py
Comment (stadlmax) [reply modulus/models/layers/fused_silu.py:22]: agreed

File: modulus/models/graphcast/graph_cast_net.py
Comment (stadlmax) [reply modulus/models/graphcast/graph_cast_net.py:202]: agreed, good suggestion

File: modulus/models/graphcast/graph_cast_net.py
Comment (stadlmax) [reply modulus/models/graphcast/graph_cast_net.py:174]: yes

File: modulus/models/gnn_layers/distributed_graph.py
Comment (stadlmax) [reply modulus/models/gnn_layers/distributed_graph.py:211]: Technically, these things are all assumed to live on the same device as device_manager.device. If they are not, routines have optional flags to indicate whether things should be broadcast/scattered/gathered/etc. I'll add more docstrings to make this clearer.

File: modulus/models/gnn_layers/distributed_graph.py
Comment (stadlmax) [reply modulus/models/gnn_layers/distributed_graph.py:300]: In relation to the comment above, I think it could be more elegant to add a more general "source-device" or a "destination device". This way, the user has more control on which device certain things should live - potentially even with CPU support.

File: modulus/models/gnn_layers/distributed_graph.py
Comment (stadlmax) [reply modulus/models/gnn_layers/distributed_graph.py:105]: Agreed. I see two steps here.
- one is simply factoring things out, I'll add this to this PR which makes addressing a few other comments easier, too
- a follow-up would be thinking about how to generalize this to other strategies, too

File: modulus/models/graphcast/graph_cast_net.py
Comment (stadlmax) [reply modulus/models/graphcast/graph_cast_net.py:134]: Unfortunately, the dataloader side hasn't really been thought about. The main use-case for this would e.g. be multi-step roll-out, i.e. when the output is partitioned, the next-steps can `expect_partitioned_input` which for now would require re-setting this flag or doing a manual partitioning before the rollout

File: modulus/models/graphcast/graph_cast_net.py
Comment (stadlmax) [reply modulus/models/graphcast/graph_cast_net.py:227]: agreed

File: modulus/models/gnn_layers/distributed_graph.py
Comment (stadlmax) [reply modulus/models/gnn_layers/distributed_graph.py:92]: agreed

File: modulus/models/gnn_layers/graph.py
Comment (stadlmax) [reply modulus/models/gnn_layers/graph.py:220]: This is something we have in many places. IIUC, this is about properly handling forward references as we can't use the class itself (here CuGraphCSC) as type when defining the class or methods.

File: modulus/models/gnn_layers/distributed_graph.py
Comment (stadlmax) [reply modulus/models/gnn_layers/distributed_graph.py:184]: I think this might be a leftover of an intermediate version which used things differently and thus this assignment needed a detach/clone as offsets was just a view of a larger global buffer. IIUC, this shouldn't be the case anymore and indeed could go away.

File: modulus/models/graphcast/graph_cast_net.py
Comment (stadlmax) [reply modulus/models/graphcast/graph_cast_net.py:133]: Currently, this is somewhat intentional actually. If no group is specified, it is just assumed that the partitioning (or in this case model parallelism) is across the global process group. 
But definitely open to suggestions here.

File: modulus/models/graphcast/graph_cast_net.py
Comment (stadlmax) [reply modulus/models/graphcast/graph_cast_net.py:132]: definitely, yes

File: modulus/models/gnn_layers/graph.py
Comment (akshaysubr) [reply modulus/models/gnn_layers/graph.py:278]: Wow, this led me down a fascinating rabbit hole and I learnt something new today! Apparently you can do C++ template style generic types in python and still have annotations and static type checking for it!

WRT this specific comment, I found the documentation for this [here](https://mypy.readthedocs.io/en/stable/generics.html#generic-methods-and-generic-self) and it seems like using `"CuGraphCSC"` would be an explicit annotation but you could also use `TypeVar` so child classes of this can also be typed appropriately. I'll leave it up to you if you want to make this change though it seems somewhat unnecessary here.

File: modulus/models/graphcast/graph_cast_net.py
Comment (akshaysubr) [reply modulus/models/graphcast/graph_cast_net.py:147]: Ah, okay. I would actually prefer if the default was to not use the global process group since the default `partition_size` is also 1. So the current defaults would work for a single GPU run, but not for non-distributed-graph training using DDP?

Comment (akshaysubr) [reply modulus/models/graphcast/graph_cast_net.py:147]: Basically, what I am getting at is that it would be good to not require the user create a single GPU process group if they just want DDP and no model parallelism.

File: modulus/models/graphcast/graph_cast_net.py
Comment (akshaysubr) [reply modulus/models/graphcast/graph_cast_net.py:148]: Okay, that makes sense. Just opened https://github.com/NVIDIA/modulus/issues/151 to track this dataloader issue.

File: modulus/models/gnn_layers/graph.py
Comment (stadlmax) [reply modulus/models/gnn_layers/graph.py:278]: Ah wow, yeah, the generic type indeed would be better in this case - despite not really necessary.

File: modulus/models/graphcast/graph_cast_net.py
Comment (stadlmax) [reply modulus/models/graphcast/graph_cast_net.py:147]: Ah, now I understand. If implemented correctly, the logic should always be
- if partition_size <= 1, the code should run in the "single-gpu" path, i.e. the process group isn't actually needed
- if partition_size > 1, the graph is assumed to be partitioned across multiple GPUs and the passed process group will be used. If passed as None, the default process group will be used. However, there are usually checks in place to assert partition_size == group_size.

I'll add a comment about the validity of process groups though and the usage of the default process group otherwise.

File: modulus/models/graphcast/graph_cast_net.py
Comment (stadlmax) [reply modulus/models/graphcast/graph_cast_net.py:148]: Thanks, definitely makes sense 👍🏽

File: modulus/models/gnn_layers/graph.py
Comment (stadlmax) [reply modulus/models/gnn_layers/graph.py:278]: Actually, it is much simpler to just use `typing.Self` as these things indeed just have a `return self`.

**Comment by mnabian** [comment]
@stadlmax could you please black format your code? I recommend using pre-commit

**Comment by stadlmax** [comment]
> @stadlmax could you please black format your code? I recommend using pre-commit

I usually do so, apparently I forgot it this time. Will do :)

File: modulus/models/gnn_layers/graph.py
Comment (akshaysubr) [reply modulus/models/gnn_layers/graph.py]: I'm not sure about using `typing.Self` since it is only supported in python 3.11 onwards according to [PEP 673](https://peps.python.org/pep-0673/). Is there a way to backport it to 3.9/3.10 which is what the NGC pytorch containers use?

File: modulus/distributed/autograd.py
Comment (akshaysubr) [inline modulus/distributed/autograd.py]: I understand why the pragma is needed here. I created a new issue to try and fix this. Can you add any other details I might have missed to this issue? https://github.com/NVIDIA/modulus/issues/165

File: modulus/models/gnn_layers/distributed_graph.py
Comment (akshaysubr) [inline modulus/models/gnn_layers/distributed_graph.py:42]: Should this be `Optional[torch.Tensor]` if default is `None`?

Comment (akshaysubr) [reply modulus/models/gnn_layers/distributed_graph.py:42]: I think generally, the defaults for many of this dataclass elements don't match the type annotations. This would cause lots of issues once we add static type checking https://github.com/NVIDIA/modulus/issues/121

File: modulus/models/gnn_layers/distributed_graph.py
Comment (akshaysubr) [inline modulus/models/gnn_layers/distributed_graph.py]: Really like this approach!

File: test/models/graphcast/utils.py
Comment (akshaysubr) [inline test/models/graphcast/utils.py]: This is a nice utility! Would be nice if you could add some docstrings so other tests can also use this.

File: test/models/meshgraphnet/test_meshgraphnet_snmg.py
Comment (akshaysubr) [inline test/models/meshgraphnet/test_meshgraphnet_snmg.py:137]: Is this not an artifact of the fact that you're using DDP with `process_group=manager.group("graph_partition")` and also creating the distributed graph using `partition_group_name="graph_partition"`? Ideally, wouldn't it be that you set up DDP with `process_group=manager.group("orthogonal_graph_partition")`?

**Review by akshaysubr — APPROVED** [review]
@stadlmax Thanks for addressing all the earlier comments. Adding some more relatively minor comments. Should be good to merge after these are addressed.

**Review by mnabian — APPROVED** [review]

File: modulus/distributed/autograd.py
Comment (stadlmax) [reply modulus/distributed/autograd.py]: No, seems good to me, thanks for opening the issue 👍🏽

File: modulus/models/gnn_layers/distributed_graph.py
Comment (stadlmax) [reply modulus/models/gnn_layers/distributed_graph.py:42]: I am actually not sure about this dataclass here.
What I want here is
- a container for the related variables defining a graph partition
- without the need to specify all of these values when initializing the object
- hence, a dataclass with mostly defaults

For tensors, you can't really define a proper default here as things like size, device, and especially dtype are only known at a later stage. I'll go with Optional for now, so technically this is not really accurate

File: test/models/graphcast/utils.py
Comment (stadlmax) [reply test/models/graphcast/utils.py]: It also would be great to have a place for these, actually.

File: test/models/meshgraphnet/test_meshgraphnet_snmg.py
Comment (stadlmax) [reply test/models/meshgraphnet/test_meshgraphnet_snmg.py:137]: IIUC, no. I would use the orthogonal process group if the the weights in each process group in `graph_partition` are not shared, right? Then you just would accumulate gradients for the same partition rank across different data-parallel groups.
Here, the model is shared and in a rather tensor-parallel setting. So the main artifact should stem from the fact that DDP by default averages the gradients based on world_size. My solution in actual code would be a custom gradient hook - though I wanted to keep the tests simpler and not deal with that.

Comment (akshaysubr) [reply test/models/meshgraphnet/test_meshgraphnet_snmg.py:137]: When the weights are shared, I was envisioning that you would instead do the all-reduce within the process group as a no-op autograd function that performs an all-reduce in the backward pass. But what you're saying is a valid approach too. This is another aspect to model parallel training that to me seems challenging to generalize. Would be good to think about this in more detail after getting this PR merged in.

**Pull request merged**
